File: minigit-3/ts/src/commands/merge.ts
// merge command - Merge branches

import * as fs from 'fs';
import * as path from 'path';
import { findRepoRoot, ensureDir, getAuthorInfo, getCommitterInfo, formatAuthor, shortSha } from '../utils';
import { readIndex, writeIndex } from '../index-file';
import {
  resolveRef,
  getHeadCommit,
  getCurrentBranch,
  updateHead
} from '../refs';
import { readObject, parseCommit, parseTree, writeCommit, writeTree, writeBlob } from '../objects';
import { mergeFiles } from '../merge-algo';
import { CommitObject, TreeEntry, IndexEntry } from '../types';

export function merge(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let noCommit = false;
  let abort = false;
  let branchName: string | null = null;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--no-commit') {
      noCommit = true;
    } else if (arg === '--abort') {
      abort = true;
    } else if (!arg.startsWith('-')) {
      branchName = arg;
    }
  }

  if (abort) {
    // Abort merge
    abortMerge(repoRoot);
    return;
  }

  if (!branchName) {
    console.error('fatal: specify a branch to merge');
    process.exit(1);
  }

  const headSha = getHeadCommit(repoRoot);
  if (!headSha) {
    console.error('fatal: HEAD does not point to a valid commit');
    process.exit(1);
  }

  const mergeSha = resolveRef(branchName, repoRoot);
  if (!mergeSha) {
    console.error(`merge: ${branchName} - not something we can merge`);
    process.exit(1);
  }

  // Check if already up to date
  if (headSha === mergeSha) {
    console.log('Already up to date.');
    return;
  }

  // Check if merge commit is ancestor of HEAD
  if (isAncestor(mergeSha, headSha, repoRoot)) {
    console.log('Already up to date.');
    return;
  }

  // Check for fast-forward
  if (isAncestor(headSha, mergeSha, repoRoot)) {
    fastForwardMerge(mergeSha, branchName, noCommit, repoRoot);
    return;
  }

  // Find merge base
  const mergeBase = findMergeBase(headSha, mergeSha, repoRoot);
  if (!mergeBase) {
    console.error('fatal: refusing to merge unrelated histories');
    process.exit(1);
  }

  // Perform three-way merge
  const result = threeWayMerge(mergeBase, headSha, mergeSha, branchName, repoRoot);

  if (result.conflicts.length > 0) {
    console.error('CONFLICT (content): Merge conflict in the following files:');
    for (const file of result.conflicts) {
      console.error(`  ${file}`);
    }
    console.error('Automatic merge failed; fix conflicts and then commit the result.');
    process.exit(1);
  }

  if (noCommit) {
    console.log(`Automatic merge went well; stopped before committing as requested`);
    return;
  }

  // Create merge commit
  const author = getAuthorInfo();
  const committer = getCommitterInfo();

  const commitObj: CommitObject = {
    tree: result.treeSha,
    parents: [headSha, mergeSha],
    author: formatAuthor(author.name, author.email, author.date),
    committer: formatAuthor(committer.name, committer.email, committer.date),
    message: `Merge branch '${branchName}'`
  };

  const commitSha = writeCommit(commitObj, repoRoot);
  updateHead(commitSha, repoRoot);

  console.log(`Merge made by the 'ort' strategy.`);
}

function fastForwardMerge(targetSha: string, branchName: string, noCommit: boolean, repoRoot: string): void {
  const headSha = getHeadCommit(repoRoot);
  const headCommit = headSha ? parseCommit(readObject(headSha, repoRoot).content) : null;
  const headFiles = new Map<string, { sha: string; mode: number }>();
  if (headCommit) {
    collectTreeFilesWithMode(headCommit.tree, '', repoRoot, headFiles);
  }

  const targetCommit = parseCommit(readObject(targetSha, repoRoot).content);
  const targetFiles = new Map<string, { sha: string; mode: number }>();
  collectTreeFilesWithMode(targetCommit.tree, '', repoRoot, targetFiles);

  // Update files in working tree
  for (const [name, entry] of targetFiles) {
    const absPath = path.join(repoRoot, name);
    ensureDir(path.dirname(absPath));
    const obj = readObject(entry.sha, repoRoot);
    fs.writeFileSync(absPath, obj.content);
    if (entry.mode === 0o100755) {
      fs.chmodSync(absPath, 0o755);
    }
  }

  // Update index
  const newEntries: IndexEntry[] = [];
  for (const [name, entry] of targetFiles) {
    const absPath = path.join(repoRoot, name);
    const stats = fs.statSync(absPath);

    newEntries.push({
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode: entry.mode,
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha: entry.sha,
      flags: Math.min(name.length, 0xfff),
      name
    });
  }
  writeIndex(newEntries, repoRoot);

  if (noCommit) {
    console.log(`Automatic merge went well; stopped before committing as requested`);
  } else {
    updateHead(targetSha, repoRoot);
    console.log(`Fast-forward`);
  }
}

function threeWayMerge(
  baseSha: string,
  oursSha: string,
  theirsSha: string,
  branchName: string,
  repoRoot: string
): { treeSha: string; conflicts: string[] } {
  const baseFiles = new Map<string, { sha: string; mode: number }>();
  const oursFiles = new Map<string, { sha: string; mode: number }>();
  const theirsFiles = new Map<string, { sha: string; mode: number }>();

  const baseCommit = parseCommit(readObject(baseSha, repoRoot).content);
  const oursCommit = parseCommit(readObject(oursSha, repoRoot).content);
  const theirsCommit = parseCommit(readObject(theirsSha, repoRoot).content);

  collectTreeFilesWithMode(baseCommit.tree, '', repoRoot, baseFiles);
  collectTreeFilesWithMode(oursCommit.tree, '', repoRoot, oursFiles);
  collectTreeFilesWithMode(theirsCommit.tree, '', repoRoot, theirsFiles);

  const allPaths = new Set([...baseFiles.keys(), ...oursFiles.keys(), ...theirsFiles.keys()]);
  const conflicts: string[] = [];
  const mergedEntries: { name: string; sha: string; mode: number }[] = [];

  for (const name of allPaths) {
    const baseEntry = baseFiles.get(name);
    const oursEntry = oursFiles.get(name);
    const theirsEntry = theirsFiles.get(name);

    const baseContent = baseEntry ? readObject(baseEntry.sha, repoRoot).content.toString() : null;
    const oursContent = oursEntry ? readObject(oursEntry.sha, repoRoot).content.toString() : null;
    const theirsContent = theirsEntry ? readObject(theirsEntry.sha, repoRoot).content.toString() : null;

    // Simple cases
    if (oursContent === theirsContent) {
      if (oursContent !== null) {
        mergedEntries.push({ name, sha: oursEntry!.sha, mode: oursEntry!.mode });
      }
      continue;
    }

    if (baseContent === oursContent && theirsContent !== null) {
      // Only theirs changed
      mergedEntries.push({ name, sha: theirsEntry!.sha, mode: theirsEntry!.mode });
      updateFile(name, theirsEntry!.sha, theirsEntry!.mode, repoRoot);
      continue;
    }

    if (baseContent === theirsContent && oursContent !== null) {
      // Only ours changed
      mergedEntries.push({ name, sha: oursEntry!.sha, mode: oursEntry!.mode });
      continue;
    }

    // Both changed - try three-way merge
    const mergeResult = mergeFiles(baseContent, oursContent, theirsContent, branchName);

    if (!mergeResult.success) {
      conflicts.push(name);
    }

    // Write merged content
    const mergedBuffer = Buffer.from(mergeResult.mergedContent || '');
    const mergedSha = writeBlob(mergedBuffer, repoRoot);
    const mode = oursEntry?.mode || theirsEntry?.mode || 0o100644;
    mergedEntries.push({ name, sha: mergedSha, mode });

    // Update working tree
    const absPath = path.join(repoRoot, name);
    ensureDir(path.dirname(absPath));
    fs.writeFileSync(absPath, mergeResult.mergedContent || '');
  }

  // Build tree from merged entries
  const treeSha = buildTree(mergedEntries, repoRoot);

  // Update index
  const newIndexEntries: IndexEntry[] = [];
  for (const entry of mergedEntries) {
    const absPath = path.join(repoRoot, entry.name);
    if (!fs.existsSync(absPath)) continue;

    const stats = fs.statSync(absPath);
    newIndexEntries.push({
      ctimeSec: Math.floor(stats.ctimeMs / 1000),
      ctimeNsec: Math.floor((stats.ctimeMs % 1000) * 1000000),
      mtimeSec: Math.floor(stats.mtimeMs / 1000),
      mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
      dev: stats.dev,
      ino: stats.ino,
      mode: entry.mode,
      uid: stats.uid,
      gid: stats.gid,
      size: stats.size,
      sha: entry.sha,
      flags: Math.min(entry.name.length, 0xfff),
      name: entry.name
    });
  }
  writeIndex(newIndexEntries, repoRoot);

  return { treeSha, conflicts };
}

function updateFile(name: string, sha: string, mode: number, repoRoot: string): void {
  const absPath = path.join(repoRoot, name);
  ensureDir(path.dirname(absPath));
  const obj = readObject(sha, repoRoot);
  fs.writeFileSync(absPath, obj.content);
  if (mode === 0o100755) {
    fs.chmodSync(absPath, 0o755);
  }
}

function buildTree(entries: { name: string; sha: string; mode: number }[], repoRoot: string): string {
  const tree = new Map<string, { name: string; sha: string; mode: number }[]>();

  for (const entry of entries) {
    const parts = entry.name.split('/');
    const topLevel = parts[0];

    if (parts.length === 1) {
      if (!tree.has('')) tree.set('', []);
      tree.get('')!.push({ name: entry.name, sha: entry.sha, mode: entry.mode });
    } else {
      if (!tree.has(topLevel)) tree.set(topLevel, []);
      tree.get(topLevel)!.push({
        name: parts.slice(1).join('/'),
        sha: entry.sha,
        mode: entry.mode
      });
    }
  }

  const treeEntries: TreeEntry[] = [];

  const rootFiles = tree.get('') || [];
  for (const file of rootFiles) {
    treeEntries.push({
      mode: file.mode.toString(8).padStart(6, '0'),
      type: 'blob',
      sha: file.sha,
      name: file.name
    });
  }

  for (const [dir, subEntries] of tree) {
    if (dir === '') continue;
    const subtreeSha = buildTree(subEntries, repoRoot);
    treeEntries.push({
      mode: '040000',
      type: 'tree',
      sha: subtreeSha,
      name: dir
    });
  }

  return writeTree(treeEntries, repoRoot);
}

function isAncestor(commitSha: string, headSha: string, repoRoot: string): boolean {
  const visited = new Set<string>();
  const queue = [headSha];

  while (queue.length > 0) {
    const sha = queue.shift()!;
    if (sha === commitSha) return true;
    if (visited.has(sha)) continue;
    visited.add(sha);

    try {
      const obj = readObject(sha, repoRoot);
      if (obj.type === 'commit') {
        const commit = parseCommit(obj.content);
        queue.push(...commit.parents);
      }
    } catch {
      // Ignore errors
    }
  }

  return false;
}

function findMergeBase(sha1: string, sha2: string, repoRoot: string): string | null {
  // Get all ancestors of sha1
  const ancestors1 = new Set<string>();
  const queue1 = [sha1];

  while (queue1.length > 0) {
    const sha = queue1.shift()!;
    if (ancestors1.has(sha)) continue;
    ancestors1.add(sha);

    try {
      const obj = readObject(sha, repoRoot);
      if (obj.type === 'commit') {
        const commit = parseCommit(obj.content);
        queue1.push(...commit.parents);
      }
    } catch {
      // Ignore errors
    }
  }

  // Find first common ancestor from sha2
  const queue2 = [sha2];
  const visited = new Set<string>();

  while (queue2.length > 0) {
    const sha = queue2.shift()!;
    if (visited.has(sha)) continue;
    visited.add(sha);

    if (ancestors1.has(sha)) {
      return sha;
    }

    try {
      const obj = readObject(sha, repoRoot);
      if (obj.type === 'commit') {
        const commit = parseCommit(obj.content);
        queue2.push(...commit.parents);
      }
    } catch {
      // Ignore errors
    }
  }

  return null;
}

function abortMerge(repoRoot: string): void {
  const headSha = getHeadCommit(repoRoot);
  if (!headSha) {
    console.error('fatal: no merge in progress');
    process.exit(1);
  }

  // Reset to HEAD
  const commit = parseCommit(readObject(headSha, repoRoot).content);
  const files = new Map<string, { sha: string; mode: number }>();
  collectTreeFilesWithMode(commit.tree, '', repoRoot, files);

  for (const [name, entry] of files) {
    const absPath = path.join(repoRoot, name);
    ensureDir(path.dirname(absPath));
    const obj = readObject(entry.sha, repoRoot);
    fs.writeFileSync(absPath, obj.content);
  }

  console.log('Merge aborted.');
}

function collectTreeFilesWithMode(treeSha: string, prefix: string, repoRoot: string, files: Map<string, { sha: string; mode: number }>): void {
  const treeObj = readObject(treeSha, repoRoot);
  const entries = parseTree(treeObj.content);

  for (const entry of entries) {
    const name = prefix ? `${prefix}/${entry.name}` : entry.name;
    if (entry.type === 'tree') {
      collectTreeFilesWithMode(entry.sha, name, repoRoot, files);
    } else {
      files.set(name, { sha: entry.sha, mode: parseInt(entry.mode, 8) });
    }
  }
}

File: minigit-3/py/commands/merge.py
"""merge command - Merge branches."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Merge branch into current branch."""
    from utils import find_repo_root
    from refs import (resolve_ref, get_head_sha, get_current_branch,
                     update_head_for_commit, read_ref)
    from objects import read_commit, create_commit, read_tree, read_blob, create_blob
    from index import read_index, write_index, add_to_index
    from merge_algo import find_common_ancestor, merge_trees

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    no_commit = False
    abort = False
    target = None

    for arg in args:
        if arg == '--no-commit':
            no_commit = True
        elif arg == '--abort':
            abort = True
        elif not arg.startswith('-'):
            target = arg

    # Handle abort
    if abort:
        merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
        if merge_head.exists():
            merge_head.unlink()
            # Reset to HEAD
            head_sha = get_head_sha(repo_root)
            if head_sha:
                from commands.checkout import update_working_tree
                update_working_tree(repo_root, head_sha)
            print("Merge aborted")
            return 0
        else:
            print("error: no merge in progress", file=sys.stderr)
            return 1

    if target is None:
        print("error: no branch specified to merge", file=sys.stderr)
        return 1

    # Resolve target
    target_sha = resolve_ref(repo_root, target)
    if target_sha is None:
        print(f"error: branch '{target}' not found", file=sys.stderr)
        return 1

    head_sha = get_head_sha(repo_root)
    if head_sha is None:
        print("error: no commits on current branch", file=sys.stderr)
        return 1

    # Check if already up-to-date
    if head_sha == target_sha:
        print("Already up to date.")
        return 0

    # Find common ancestor
    base_sha = find_common_ancestor(repo_root, head_sha, target_sha)

    # Check for fast-forward
    if base_sha == head_sha:
        if no_commit:
            # For --no-commit with fast-forward, stage the changes but don't move HEAD
            # Update working tree and index to match target
            from commands.checkout import update_working_tree
            update_working_tree(repo_root, target_sha)

            # Write MERGE_HEAD for later commit
            merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
            merge_head.write_text(target_sha + '\n')

            print(f"Automatic merge went well; stopped before committing as requested")
            return 0
        else:
            # Fast-forward merge
            update_head_for_commit(repo_root, target_sha)

            # Update working tree
            from commands.checkout import update_working_tree
            update_working_tree(repo_root, target_sha)

            print(f"Fast-forward merge to {target_sha[:7]}")
            return 0

    # Check if already merged
    if base_sha == target_sha:
        print("Already up to date.")
        return 0

    # Three-way merge
    head_commit = read_commit(repo_root, head_sha)
    target_commit = read_commit(repo_root, target_sha)

    base_tree = None
    if base_sha:
        base_commit = read_commit(repo_root, base_sha)
        base_tree = base_commit['tree']

    # Get current branch name for labels
    current_branch = get_current_branch(repo_root) or 'HEAD'

    # Merge trees
    merged, has_conflicts = merge_trees(
        repo_root,
        base_tree,
        head_commit['tree'],
        target_commit['tree'],
        current_branch,
        target
    )

    # Update working tree and index
    new_entries = []

    def get_tree_files(tree_sha: str, prefix: str = '') -> dict:
        files = {}
        try:
            tree_entries = read_tree(repo_root, tree_sha)
            for entry in tree_entries:
                path = f"{prefix}{entry.name}" if prefix else entry.name
                if entry.mode == '40000':
                    files.update(get_tree_files(entry.sha, path + '/'))
                else:
                    files[path] = {'sha': entry.sha, 'mode': entry.mode}
        except:
            pass
        return files

    # Start with files from HEAD
    head_files = get_tree_files(head_commit['tree'])

    for path, info in merged.items():
        full_path = repo_root / path
        full_path.parent.mkdir(parents=True, exist_ok=True)

        if 'conflict' in info and info['conflict']:
            # Write conflict content
            if 'content' in info:
                full_path.write_text(info['content'])
            else:
                # Binary conflict - keep ours
                if 'ours_sha' in info:
                    content = read_blob(repo_root, info['ours_sha'])
                    full_path.write_bytes(content)

            # Don't add to index (leave as conflicted)
            # But we need to keep track of it somehow
            stat = full_path.stat()
            new_entries.append({
                'path': path,
                'sha': head_files.get(path, {}).get('sha', '0' * 40),
                'mode': info.get('mode', '100644'),
                'ctime': (int(stat.st_ctime), int((stat.st_ctime % 1) * 1e9)),
                'mtime': (int(stat.st_mtime), int((stat.st_mtime % 1) * 1e9)),
                'dev': stat.st_dev,
                'ino': stat.st_ino,
                'uid': stat.st_uid,
                'gid': stat.st_gid,
                'size': stat.st_size,
                'flags': min(len(path), 0xFFF)
            })
        else:
            # Write merged content
            content = read_blob(repo_root, info['sha'])
            full_path.write_bytes(content)

            if info['mode'] == '100755':
                full_path.chmod(0o755)

            stat = full_path.stat()
            new_entries.append({
                'path': path,
                'sha': info['sha'],
                'mode': info['mode'],
                'ctime': (int(stat.st_ctime), int((stat.st_ctime % 1) * 1e9)),
                'mtime': (int(stat.st_mtime), int((stat.st_mtime % 1) * 1e9)),
                'dev': stat.st_dev,
                'ino': stat.st_ino,
                'uid': stat.st_uid,
                'gid': stat.st_gid,
                'size': stat.st_size,
                'flags': min(len(path), 0xFFF)
            })

    # Remove files not in merged result
    for path in head_files:
        if path not in merged:
            full_path = repo_root / path
            if full_path.exists():
                full_path.unlink()

    write_index(repo_root, new_entries)

    if has_conflicts:
        # Write MERGE_HEAD for --abort
        merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
        merge_head.write_text(target_sha + '\n')

        print(f"Automatic merge failed; fix conflicts and then commit the result.")
        return 1

    if no_commit:
        # Write MERGE_HEAD for later commit
        merge_head = repo_root / '.minigit' / 'MERGE_HEAD'
        merge_head.write_text(target_sha + '\n')

        print(f"Automatic merge went well; stopped before committing as requested")
        return 0

    # Create merge commit
    from objects import build_tree_from_index

    tree_sha = build_tree_from_index(repo_root, new_entries)
    message = f"Merge branch '{target}'"

    commit_sha = create_commit(
        repo_root, tree_sha, message,
        parents=[head_sha, target_sha]
    )

    update_head_for_commit(repo_root, commit_sha)

    print(f"Merge made by the 'recursive' strategy.")
    return 0
